File: 14501/yeolyi.py
from calendar import c


N = int(input())
arr = []
for i in range(N):
    arr.append(list(map(int, input().split(' '))))

cache = {}

def solve(start):
    if start in cache:
        return cache[start]
    if N <= start:
        return 0
    ret = 0
    for i in range(start, N):
        length = arr[i][0]
        nextDay = i + length
        if N < nextDay:
            continue
        cost = arr[i][1]
        val = cost + solve(nextDay)
        ret = max(ret, val)
    cache[start] = ret
    return ret

print(solve(0))

File: 2116/yeolyi.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const arr = input.slice(1).map(x => x.split(' ').map(Number));
const matchingIndex = [5, 3, 4, 1, 2, 0];

function getSideMax(up, bottom) {
    if (up === 6) {
        return bottom === 5 ? 4 : 5;
    } else if (bottom === 6) {
        return up === 5 ? 4 : 5;
    }
    return 6;
}

function solve(bottomVal, dice) {
    const bottomIdx = dice.indexOf(bottomVal);
    const upIdx = matchingIndex[bottomIdx];
    const upVal = dice[upIdx];

    const sideMax = getSideMax(upVal, bottomVal);
    return [upVal, sideMax];
}

let result = 0;

for (let _upVal = 1; _upVal < 7; _upVal++) {
    let upVal = _upVal;
    let total = 0;
    for (let diceIdx=0; diceIdx < arr.length; diceIdx++) {
        if (result >= total+6*(arr.length-diceIdx)) { break; }
        const dice = arr[diceIdx];
        const [nextUpVal, sideMax] = solve(upVal, dice);
        upVal = nextUpVal;
        total += sideMax;
    }
    result = Math.max(result, total);
}

console.log(result);

